pkg/adapterManager/manager.go

```diff
@@ -95,32 +99,97 @@ func newCacheKey(kind aspect.Kind, cfg *config.Combined) (*cacheKey, error) {
 }
 
 // NewManager creates a new adapterManager.
-func NewManager(builders []adapter.RegisterFn, managers aspect.ManagerInventory, exp expr.Evaluator) *Manager {
+func NewManager(builders []adapter.RegisterFn, managers aspect.ManagerInventory, exp expr.Evaluator, sequential bool) *Manager {
 	mm, am := ProcessBindings(managers)
-	return newManager(newRegistry(builders), mm, exp, am)
+	return newManager(newRegistry(builders), mm, exp, am, sequential)
 }
 
-// newManager
-func newManager(r builderFinder, m map[aspect.Kind]aspect.Manager, exp expr.Evaluator, am map[aspect.APIMethod]config.AspectSet) *Manager {
+func newManager(r builderFinder, m map[aspect.Kind]aspect.Manager, exp expr.Evaluator, am map[aspect.APIMethod]config.AspectSet, sequential bool) *Manager {
 	return &Manager{
 		builders:    r,
 		managers:    m,
 		mapper:      exp,
 		methodMap:   am,
 		aspectCache: make(map[cacheKey]aspect.Wrapper),
+		sequential:  sequential,
 	}
 }
 
+// Shutdown gracefully drains the manager's worker pool.
+func (m *Manager) Shutdown() {
+	m.wg.Wait()
+}
+
 // Execute iterates over cfgs and performs the actions described by the combined config using the attribute bag on each config.
 func (m *Manager) Execute(ctx context.Context, cfgs []*config.Combined, attrs attribute.Bag, ma aspect.APIMethodArgs) aspect.Output {
-	// TODO: pool these arrays, we'll be making many and len(cfg) is constant for the life of the configuration.
-	results := make([]result, len(cfgs))
-	for i, cfg := range cfgs {
-		results[i] = result{cfg, m.execute(ctx, cfg, attrs, ma)}
+	numCfgs := len(cfgs)
+
+	// TODO: look into pooling both result array and channel, they're created per-request and are constant size for cfg lifetime.
+	results := make([]result, numCfgs)
+	resultChan := make(chan result, numCfgs)
+
+	if m.sequential {
+		for i, cfg := range cfgs {
+			results[i] = result{cfg, m.execute(ctx, cfg, attrs, ma)}
+		}
+	} else {
+		// schedule all the work that needs to happen
+		for _, cfg := range cfgs {
+			m.wg.Add(1)
+			pool.ScheduleWork(func() {
```

Comment:
I can easily change ScheduleWork to deal with blocking/non-blocking:

  func ScheduleWork(fn WorkFunc, bool nonBlocking) bool

So you can request a non-blocking version and you get a boolean back telling whether or not the work was scheduled. 

This would make it easy to fix the logic you point out. However, the reason I didn't do this is 'cause I thought it didn't need to be done.

If the set of worker goroutines is busy and the channel queue fills up, then ScheduleWork will block. In the case of the adapter manager, that means the adapter manager won't be able to make forward progress until some work completes. Everything will eventually back up and the goroutine that's pumping stuff from the grpc stream will also block. This will eventually block at the TCP level, which will stop traffic coming in.

It's actually beneficial to have backpressure like this propagate all the way down to the network and out to clients. As compared to having explicit logic to detect we're saturated and returning failures to the callers explicitly.